Fix #5127 Memory corruption with the VM

maximedenes opened the pull request:
The bytecode interpreter ensures that the stack space available at some
points is above a static threshold. However, arbitrary large stack space
can be needed between two check points, leading to segmentation faults
in some cases.

We track the use of stack space at compilation time and add
an instruction to ensure greater stack capacity when required. This is
inspired from OCaml's PR#339 and PR#7168.

maximedenes commented:
@silene could you have a look to tell us if this solves all the problems you had in mind? Thanks!

maximedenes commented:
Wow, there seems to be a big flaw in this patch, so wait a bit before reviewing.

maximedenes commented:
Ok, problem fixed, I think it is ready for review now.

JasonGross commented:
What performance impact, if any, does this have on the vm?

maximedenes commented:
It should have essentially no cost, because the instruction making the stack grow is added only in the cases when you could do a segfault previously (according to a static criterion).

JasonGross commented:
> the instruction making the stack grow is added only in the cases when you could do a segfault previously (according to a static criterion).

Is this an exception to that rule, or am I misreading the code?
```ocaml
      (* The fun code of [vf] will make sure we have enough stack, so we put 0
         here. *)
      push_vstack varray 0;
```

maximedenes commented:
No, these checks are necessary. A value of "0" means that the interpreter will check that the stack free space is still above the usual threshold, which the rest of the code assumes. If you look at the code of push_vstack, it was simply missing checks before.

silene commented on kernel/byterun/coq_interp.c:
```diff
@@ -422,6 +429,17 @@ value coq_interprete
+          coq_sp = sp;
+          realloc_coq_stack(size * sizeof(value));
+          sp = coq_sp;
+        }
```
This seems incorrect. We might end up with no redzone once the stack has been extended and filled, if we are unlucky. Why is this code not just `CHECK_STACK(size)` so as to ensure the redzone is accounted for?

silene replied:
No, stack doubling is irrelevant here (though it does make the issue less likely to occur). Imagine your stack has size N and it is almost completely filled. Imagine the current function needs about N stack slots before calling another function (and as I explained in the bug-report, this other function does not even need to have arity N for that). So `ENSURESTACKCAPACITY` causes the stack to be reallocated to `2*N` slots. Then the current function tries to perform the call and it reallocates the stack to `4*N` slots because the threshold has been reached. So we are paying two reallocations in a row while we could have reasonably guessed the second reallocation was mandatory because we were about to reach the threshold. The only case where taking the threshold into account in `ENSURESTACKCAPACITY` is a pessimization is when the function does not perform a call (while it required a huge stack).

silene commented on kernel/byterun/coq_interp.c:
```diff
@@ -76,6 +76,14 @@ sp is a local copy of the global variable extern_sp. */
+#define CHECK_STACK(num_args) {                                                \
+if (sp - num_args < coq_stack_threshold) {                                     \
+  coq_sp = sp;                                                                 \
+  realloc_coq_stack(num_args * sizeof(value) + Coq_stack_threshold);           \
```
This seems dubious. `realloc_coq_stack` takes as argument the number of words to allocate. So `num_args` should not be multiplied by `sizeof(value)`.

maximedenes replied:
Yes, I was a bit confused about this too. So the pre-existing code of `check_stack:` (see below) was also wrong, right? It was calling realloc_coq_stack with Coq_stack_threshold. We should give more explicit names to Coq_stack_threshold vs coq_stack_threshold to avoid confusion.

silene commented on kernel/cbytegen.ml:
```diff
@@ -684,12 +730,13 @@ let rec compile_constr reloc c sz cont =
+	fun_code := [Ksequence(add_grab (arity+1) lbl cont,!fun_code)];
       done;
       let fv = !rfv in
       compile_fv reloc fv.fv_rev sz
```
`Kclosurecofix` needs `2 + ndef` extra stack slots, which are not accounted for, if I am not mistaken.

silene commented on kernel/cbytegen.ml:
```diff
@@ -535,12 +574,13 @@ let compile_fv_elem reloc fv sz cont =
   match l with
-  | [] -> cont
+  | [] -> set_max_stack_size sz; cont
   | [fvn] -> compile_fv_elem reloc fvn sz cont
```
Since there is a special case for the length-1 list, the length-0 list will almost never be triggered (and when it is, you don't need any extra stack space), so the code seems plain wrong here. It should have been something like

    | [] -> cont
    | [fvn] -> set_max_stack_size (sz + 1); compile_fv_elem reloc fvn sz cont

In case you wonder why I suggest `+ 1` even though there is no push in that branch, that is because there is an implicit push in all the opcodes that prepare a closure:

    if (nvars > 0) *--sp = accu;

maximedenes commented:
I think we addressed all remarks. A big thank to @silene for his very thorough review.

Will merge if tests go fine: https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/546/console
